plugins/network-elements/nuage-vsp/src/com/cloud/network/element/NuageVspElement.java

```diff
@@ -205,9 +254,79 @@ public boolean implement(Network network, NetworkOffering offering, DeployDestin
+            isL3Network = true;
+        } else {
+            isL2Network = true;
+        }
+
+        String preConfiguredDomainTemplateName = NuageVspUtil.getPreConfiguredDomainTemplateName(_configDao, network, offering);
+        List<FirewallRuleVO> firewallIngressRulesToApply = getFirewallRulesToApply(network.getId(), FirewallRule.TrafficType.Ingress);
+        List<FirewallRuleVO> firewallEgressRulesToApply = getFirewallRulesToApply(network.getId(), FirewallRule.TrafficType.Egress);
+
+        List<Map<String, Object>> ingressFirewallRules = Lists.transform(firewallIngressRulesToApply, new Function<FirewallRuleVO, Map<String, Object>>() {
+            @Override
+            public Map<String, Object> apply(FirewallRuleVO firewallRuleVO) {
+                return getACLRuleDetails(firewallRuleVO, egressDefaultPolicy);
+            }
+        });
+
+        List<Map<String, Object>> egressFirewallRules = Lists.transform(firewallEgressRulesToApply, new Function<FirewallRuleVO, Map<String, Object>>() {
+            @Override
+            public Map<String, Object> apply(FirewallRuleVO firewallRuleVO) {
+                return getACLRuleDetails(firewallRuleVO, egressDefaultPolicy);
+            }
+        });
+
+        List<IPAddressVO> ips = _ipAddressDao.listStaticNatPublicIps(network.getId());
+        List<String> acsFipUuid = new ArrayList<String>();
+        for (IPAddressVO ip : ips) {
+            acsFipUuid.add(ip.getUuid());
+        }
+
+        try {
+            HostVO nuageVspHost = getNuageVspHost(network.getPhysicalNetworkId());
+            ImplementVspCommand.Builder cmdBuilder = new ImplementVspCommand.Builder().networkId(network.getId()).networkDomainUuid(networkDomain.getUuid())
+                    .networkUuid(network.getUuid()).networkName(network.getName()).vpcOrSubnetUuid(subnetUuid).isL2Network(isL2Network).isL3Network(isL3Network)
+                    .isVpc(false).isShared(isShared).domainTemplateName(preConfiguredDomainTemplateName).isFirewallServiceSupported(isFirewallServiceSupported)
+                    .dnsServers(dnsServers).ingressFirewallRules(ingressFirewallRules).egressFirewallRules(egressFirewallRules).acsFipUuid(acsFipUuid)
+                    .egressDefaultPolicy(egressDefaultPolicy);
+            ImplementVspAnswer answer = (ImplementVspAnswer) _agentMgr.easySend(nuageVspHost.getId(), cmdBuilder.build());
+            if (answer == null || !answer.getResult()) {
+                s_logger.error("ImplementVspCommand for network " + network.getUuid() + " failed");
+                if ((null != answer) && (null != answer.getDetails())) {
+                    throw new ResourceUnavailableException(answer.getDetails(), Network.class, network.getId());
+                }
+            }
+        } catch (Exception e) {
+            s_logger.warn("Failed to implement network in Vsp", e);
```

Comment:
Why is this message being logged to ``WARN`` instead of ``ERROR``.  The behavior indicates that this condition halts operation.  Also, please add context information about the VSP to help admins determine instance encountered a problem.